The Swift compiler cannot be built with clang 15 on linux/Android

**Describe the bug**
I tried updating to building with clang 15 and could no longer build this repo, with errors like this:
```
[3/1065][  0%][1.958s] Building CXX object lib/SwiftReflection/CMakeFiles/swiftReflection.dir/__/__/stdlib/public/Reflection/TypeLowering.cpp.o
FAILED: lib/SwiftReflection/CMakeFiles/swiftReflection.dir/__/__/stdlib/public/Reflection/TypeLowering.cpp.o 
/home/butta/clang-15.0.2/bin/clang++ -DCMARK_STATIC_DEFINE -DGTEST_HAS_RTTI=0 -DNDEBUG -DSWIFT_THREADING_LINUX -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/butta/build/Ninja-Release/swift-linux-x86_64/lib/SwiftReflection -I/home/butta/swift/lib/SwiftReflection -I/home/butta/swift/stdlib/public/SwiftShims -I/home/butta/build/Ninja-Release/swift-linux-x86_64/include -I/home/butta/swift/include -I/home/butta/llvm-project/llvm/include -I/home/butta/build/Ninja-Release/llvm-linux-x86_64/include -I/home/butta/llvm-project/clang/include -I/home/butta/build/Ninja-Release/llvm-linux-x86_64/tools/clang/include -I/home/butta/cmark/src/include -I/home/butta/build/Ninja-Release/cmark-linux-x86_64/src -I/home/butta/swift-corelibs-libdispatch/src/BlocksRuntime -I/home/butta/swift-corelibs-libdispatch -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-class-memaccess -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Werror=switch -Wdocumentation -Wimplicit-fallthrough -Wunreachable-code -Woverloaded-virtual -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -O3 -DNDEBUG  -fno-exceptions -fno-rtti -Werror=gnu -Werror=c++98-compat-extra-semi -target x86_64-unknown-linux-gnu -O2 -momit-leaf-frame-pointer -g0 -march=core2 -DSWIFT_ENABLE_REFLECTION -std=c++14 -MD -MT lib/SwiftReflection/CMakeFiles/swiftReflection.dir/__/__/stdlib/public/Reflection/TypeLowering.cpp.o -MF lib/SwiftReflection/CMakeFiles/swiftReflection.dir/__/__/stdlib/public/Reflection/TypeLowering.cpp.o.d -o lib/SwiftReflection/CMakeFiles/swiftReflection.dir/__/__/stdlib/public/Reflection/TypeLowering.cpp.o -c /home/butta/swift/stdlib/public/Reflection/TypeLowering.cpp
In file included from /home/butta/swift/stdlib/public/Reflection/TypeLowering.cpp:26:
In file included from /home/butta/swift/include/swift/Reflection/TypeLowering.h:24:
In file included from /home/butta/swift/include/swift/Remote/MetadataReader.h:20:
In file included from /home/butta/swift/include/swift/Runtime/Metadata.h:20:
In file included from /home/butta/swift/include/swift/ABI/Metadata.h:43:
In file included from /home/butta/swift/stdlib/public/SwiftShims/swift/shims/HeapObject.h:15:
In file included from /home/butta/swift/stdlib/public/SwiftShims/swift/shims/RefCount.h:36:
/home/butta/swift/include/swift/Basic/type_traits.h:49:29: warning: builtin __has_trivial_constructor is deprecated; use __is_trivially_constructible instead [-Wdeprecated-builtins]
  static const bool value = __has_trivial_constructor(T);
                            ^
/home/butta/swift/include/swift/Basic/type_traits.h:61:29: warning: builtin __has_trivial_destructor is deprecated; use __is_trivially_destructible instead [-Wdeprecated-builtins]
  static const bool value = __has_trivial_destructor(T);
                            ^
In file included from /home/butta/swift/stdlib/public/Reflection/TypeLowering.cpp:26:
In file included from /home/butta/swift/include/swift/Reflection/TypeLowering.h:24:
In file included from /home/butta/swift/include/swift/Remote/MetadataReader.h:20:
In file included from /home/butta/swift/include/swift/Runtime/Metadata.h:20:
In file included from /home/butta/swift/include/swift/ABI/Metadata.h:43:
In file included from /home/butta/swift/stdlib/public/SwiftShims/swift/shims/HeapObject.h:15:
In file included from /home/butta/swift/stdlib/public/SwiftShims/swift/shims/RefCount.h:40:
/home/butta/swift/include/swift/Runtime/Heap.h:32:1: error: an attribute list cannot appear here
SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/butta/swift/stdlib/public/SwiftShims/swift/shims/Visibility.h:240:31: note: expanded from macro 'SWIFT_RETURNS_NONNULL'
#define SWIFT_RETURNS_NONNULL [[gnu::returns_nonnull]]
                              ^
In file included from /home/butta/swift/stdlib/public/Reflection/TypeLowering.cpp:26:
In file included from /home/butta/swift/include/swift/Reflection/TypeLowering.h:24:
In file included from /home/butta/swift/include/swift/Remote/MetadataReader.h:20:
/home/butta/swift/include/swift/Runtime/Metadata.h:316:1: error: an attribute list cannot appear here
SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/butta/swift/stdlib/public/SwiftShims/swift/shims/Visibility.h:240:31: note: expanded from macro 'SWIFT_RETURNS_NONNULL'
#define SWIFT_RETURNS_NONNULL [[gnu::returns_nonnull]]
                              ^
In file included from /home/butta/swift/stdlib/public/Reflection/TypeLowering.cpp:26:
In file included from /home/butta/swift/include/swift/Reflection/TypeLowering.h:24:
In file included from /home/butta/swift/include/swift/Remote/MetadataReader.h:20:
/home/butta/swift/include/swift/Runtime/Metadata.h:325:1: error: an attribute list cannot appear here
SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/butta/swift/stdlib/public/SwiftShims/swift/shims/Visibility.h:240:31: note: expanded from macro 'SWIFT_RETURNS_NONNULL'
#define SWIFT_RETURNS_NONNULL [[gnu::returns_nonnull]]
                              ^
In file included from /home/butta/swift/stdlib/public/Reflection/TypeLowering.cpp:26:
In file included from /home/butta/swift/include/swift/Reflection/TypeLowering.h:24:
In file included from /home/butta/swift/include/swift/Remote/MetadataReader.h:30:
/home/butta/swift/include/swift/Runtime/HeapObject.h:64:1: error: an attribute list cannot appear here
SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/butta/swift/stdlib/public/SwiftShims/swift/shims/Visibility.h:240:31: note: expanded from macro 'SWIFT_RETURNS_NONNULL'
#define SWIFT_RETURNS_NONNULL [[gnu::returns_nonnull]]
                              ^
In file included from /home/butta/swift/stdlib/public/Reflection/TypeLowering.cpp:26:
In file included from /home/butta/swift/include/swift/Reflection/TypeLowering.h:24:
In file included from /home/butta/swift/include/swift/Remote/MetadataReader.h:30:
/home/butta/swift/include/swift/Runtime/HeapObject.h:120:1: error: an attribute list cannot appear here
SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/butta/swift/stdlib/public/SwiftShims/swift/shims/Visibility.h:240:31: note: expanded from macro 'SWIFT_RETURNS_NONNULL'
#define SWIFT_RETURNS_NONNULL [[gnu::returns_nonnull]]
                              ^
2 warnings and 5 errors generated.
ninja: build stopped: subcommand failed.
```
I first saw this when building trunk natively on Android AArch64.

**Steps To Reproduce**
1. Since Ubuntu 20.04 x86_64 doesn't ship a system package more recent than clang 12, I first downloaded [the official 15.0.2 build for linux x86_64, `clang+llvm-15.0.2-x86_64-unknown-linux-gnu-rhel86.tar.xz`](https://github.com/llvm/llvm-project/releases/tag/llvmorg-15.0.2), unpacked it, and renamed it to `clang-15.0.2/`.
2. `./swift/utils/build-script -RA -j9 --skip-early-swift-driver --host-cc=/home/butta/clang-15.0.2/bin/clang --host-cxx=/home/butta/clang-15.0.2/bin/clang++`

**Expected behavior**
No problems, as with clang 14

**Environment**
 - OS: Ubuntu 20.04 x86_64 and Android 12 AArch64

**Additional context**
This patch works around the problem on Android, so presumably these were the two C++ attributes recently enabled by clang 15:
```
diff --git a/stdlib/public/SwiftShims/swift/shims/Visibility.h b/stdlib/public/SwiftShims/swift/shims/Visibility.h
index 84d22d20fa2..c18282bd3ff 100644
--- a/stdlib/public/SwiftShims/swift/shims/Visibility.h
+++ b/stdlib/public/SwiftShims/swift/shims/Visibility.h
@@ -228,16 +228,18 @@
 #define SWIFT_FALLTHROUGH
 #endif
 
+//[[clang::warn_unused_result]]
 #if __cplusplus > 201402l && __has_cpp_attribute(nodiscard)
 #define SWIFT_NODISCARD [[nodiscard]]
 #elif __has_cpp_attribute(clang::warn_unused_result)
-#define SWIFT_NODISCARD [[clang::warn_unused_result]]
+#define SWIFT_NODISCARD
 #else
 #define SWIFT_NODISCARD
 #endif
 
+//[[gnu::returns_nonnull]]
 #if __has_cpp_attribute(gnu::returns_nonnull)
-#define SWIFT_RETURNS_NONNULL [[gnu::returns_nonnull]]
+#define SWIFT_RETURNS_NONNULL
 #elif defined(_MSC_VER) && defined(_Ret_notnull_)
 #define SWIFT_RETURNS_NONNULL _Ret_notnull_
 #else
```

--- a/include/swift/Runtime/Config.h
+++ b/include/swift/Runtime/Config.h
@@ -168,6 +168,14 @@ extern uintptr_t __COMPATIBILITY_LIBRARIES_CANNOT_CHECK_THE_IS_SWIFT_BIT_DIRECTL
 // Bring in visibility attribute macros
 #include "swift/shims/Visibility.h"
 
+// Temporary definitions to allow compilation on clang-15.
+#if defined(__cplusplus)
+#define SWIFT_EXTERN_C extern "C"
+#else
+#define SWIFT_EXTERN_C
+#endif
+#define SWIFT_RUNTIME_EXPORT_ATTRIBUTE SWIFT_EXPORT_FROM_ATTRIBUTE(swiftCore)
+
 // Define mappings for calling conventions.
 
 // Annotation for specifying a calling convention of

--- a/include/swift/Runtime/Heap.h
+++ b/include/swift/Runtime/Heap.h
@@ -29,7 +29,7 @@ namespace swift {
 // Never returns nil. The returned memory is uninitialized. 
 //
 // An "alignment mask" is just the alignment (a power of 2) minus 1.
-SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT
+SWIFT_EXTERN_C SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT_ATTRIBUTE
 void *swift_slowAlloc(size_t bytes, size_t alignMask);
 
 // If the caller cannot promise to zero the object during destruction,

--- a/include/swift/Runtime/HeapObject.h
+++ b/include/swift/Runtime/HeapObject.h
@@ -61,7 +61,7 @@ struct OpaqueValue;
 ///
 /// POSSIBILITIES: The argument order is fair game.  It may be useful
 /// to have a variant which guarantees zero-initialized memory.
-SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT
+SWIFT_EXTERN_C SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT_ATTRIBUTE
 HeapObject *swift_allocObject(HeapMetadata const *metadata,
                               size_t requiredSize,
                               size_t requiredAlignmentMask);
@@ -117,7 +117,7 @@ BoxPair swift_makeBoxUnique(OpaqueValue *buffer, Metadata const *type,
                                     size_t alignMask);
 
 /// Returns the address of a heap object representing all empty box types.
-SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT
+SWIFT_EXTERN_C SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT_ATTRIBUTE
 HeapObject* swift_allocEmptyBox();
 
 /// Atomically increments the retain count of an object.

--- a/include/swift/Runtime/Metadata.h
+++ b/include/swift/Runtime/Metadata.h
@@ -313,7 +313,7 @@ swift_getGenericMetadata(MetadataRequest request,
 ///   - installing new v-table entries and overrides; and
 ///   - registering the class with the runtime under ObjC interop.
 /// Most of this work can be achieved by calling swift_initClassMetadata.
-SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT
+SWIFT_EXTERN_C SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT_ATTRIBUTE
 ClassMetadata *
 swift_allocateGenericClassMetadata(const ClassDescriptor *description,
                                    const void *arguments,
@@ -322,7 +322,7 @@ swift_allocateGenericClassMetadata(const ClassDescriptor *description,
 /// Allocate a generic value metadata object.  This is intended to be
 /// called by the metadata instantiation function of a generic struct or
 /// enum.
-SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT
+SWIFT_EXTERN_C SWIFT_RETURNS_NONNULL SWIFT_NODISCARD SWIFT_RUNTIME_EXPORT_ATTRIBUTE
 ValueMetadata *
 swift_allocateGenericValueMetadata(const ValueTypeDescriptor *description,
                                    const void *arguments,

